packages/react/src/components/FocusTrapZone/FocusTrapZone.tsx

```diff
@@ -97,284 +92,230 @@ export const FocusTrapZone: React.FunctionComponent<IFocusTrapZoneProps> & {
+      typeof elementToFocusOnDismiss?.focus === 'function' &&
+      // only restore focus if the current focused element is within the FTZ, or if nothing is focused
+      (elementContains(root.current, activeElement) || activeElement === doc.body)
+    ) {
+      focusElementAsync(elementToFocusOnDismiss);
+    }
+  });
 
-  const forceClickInTrap = React.useCallback(
-    (ev: MouseEvent): void => {
-      if (disabled) {
-        return;
-      }
-      if (FocusTrapZone.focusStack.length && id === FocusTrapZone.focusStack[FocusTrapZone.focusStack.length - 1]) {
-        const clickedElement = ev.target as HTMLElement;
-        if (clickedElement && !elementContains(root.current, clickedElement)) {
-          focus();
-          internalState.hasFocus = true; // set focus here since we stop event propagation
-          ev.preventDefault();
-          ev.stopPropagation();
-        }
+  /** Called in window event handlers. (useEventCallback ensures latest prop values are used.) */
+  const forceFocusOrClickInTrap = useEventCallback((ev: FocusEvent | MouseEvent): void => {
+    // be sure to use the latest values here
+    if (disabled) {
+      return;
+    }
+    if (internalState.focusStackId === FocusTrapZone.focusStack!.slice(-1)[0]) {
+      const targetElement = ev.target as HTMLElement | null;
+      if (targetElement && !elementContains(root.current, targetElement)) {
+        focusFTZ();
+        internalState.hasFocus = true; // set focus here since we stop event propagation
+        ev.preventDefault();
+        ev.stopPropagation();
       }
-    },
-    [disabled, id, focus, internalState],
-  );
-
-  const updateEventHandlers = React.useCallback((): void => {
-    if (forceFocusInsideTrap && !internalState.disposeFocusHandler) {
-      internalState.disposeFocusHandler = on(window, 'focus', forceFocusInTrap, true);
-    } else if (!forceFocusInsideTrap && internalState.disposeFocusHandler) {
-      internalState.disposeFocusHandler();
-      internalState.disposeFocusHandler = undefined;
     }
+  });
 
-    if (!isClickableOutsideFocusTrap && !internalState.disposeClickHandler) {
-      internalState.disposeClickHandler = on(window, 'click', forceClickInTrap, true);
-    } else if (isClickableOutsideFocusTrap && internalState.disposeClickHandler) {
-      internalState.disposeClickHandler();
-      internalState.disposeClickHandler = undefined;
+  // Update window event handlers when relevant props change
+  React.useEffect(() => {
+    const disposables: Array<() => void> = [];
+
+    if (forceFocusInsideTrap) {
+      disposables.push(on(window, 'focus', forceFocusOrClickInTrap, true));
+    }
+    if (!isClickableOutsideFocusTrap) {
+      disposables.push(on(window, 'click', forceFocusOrClickInTrap, true));
     }
-  }, [forceClickInTrap, forceFocusInTrap, forceFocusInsideTrap, isClickableOutsideFocusTrap, internalState]);
 
-  // Updates eventHandlers and cleans up focusStack when the component unmounts.
```

Comment:
Everything that was in this effect is now handled by other effects' cleanup (or no longer relevant)